### ISSUE TITLE
VNAV: Speed-Bug issue during deceleration before 10.000 feet

### ISSUE THREAD

**Issue by oMrSmith** [mSparks43/747-400#548]
- [ ] Bug: VNAV dosn't bug for 240 knots
until it has descended down to 10.000 ft, which is too late.
It should bug for 240 knots
_as soon as_ the deceleration has started.

Otherwise the thrust mode SPD will 
hold the CRZ SPD until the plane has 
descended down to 10.000 ft which
is counterproductive:

![image](https://user-images.githubusercontent.com/78870770/118466591-4a0bbc00-b703-11eb-92cb-155325763db4.png)

https://youtu.be/SwK09XdCZAQ?t=21

Happens on low cruzing altitude 11.000 feet...

**Comment by mSparks43** [comment]
I have no idea what/how a fl110 cruise altitude should work. is there some reference for this or is it just from testing?

**Comment by oMrSmith** [comment]
I think cruzing on fl110 is all fine. Just if you pass the T/D the A/T won't IDLE the thrust,
so that the plane won't slow down...

**Comment by mSparks43** [comment]
cruising is fine, but transition from stable cruise to stable descent takes about 1500 feet, and trying to bleed off the speed before that would blow miles past the descent path.

So the question is, does VNAV do this or the pilots?

**Comment by oMrSmith** [comment]
I believe this: After passing the T/D the plane should simply initiate an IDLE descend
with the A/T switching from SPD to IDLE. With the descend rate of only 500 fpm
the plane would then immediately start to decelerate. If its still faster than 250 knots
when reaching 10.000 ft the pilot can intervene.   

In my testflight however the plane started the descend but
the A/T would keep SPD mode active which worked
against the deceleration and delayed it.

**Comment by mSparks43** [comment]
Or.
The proper procedure when flying so low with a speed restriction near and just below ToD is to trigger the descent up to 50nm before ToD by dialing in a lower alt and pressing the alt button, giving a much flatter descent profile and more time to burn off the speed.

Or.
when flying so low with a speed restriction near and just below ToD the ToD is increased automatically to account for the need to burn off speed much faster than is normal

The problem with this is without some reference its impossible to know one way or another.

**Comment by oMrSmith** [comment]
I could test this with the 747 by SSG but it's not considered to be all too realistic...

**Comment by oMrSmith** [comment]
But there's another question that interests me:
Should VNAV bug for 250 when reaching 10.000 feet?'
or
Should VNAV already bug for 250 when starting the deceleration?

**Comment by oMrSmith** [comment]
Ok. Just tested how it works in the SSG:

![image](https://user-images.githubusercontent.com/78870770/118413163-367a3a00-b69e-11eb-8571-7596cf24672d.png)

1. When the descend starts, the A/T immediately switches from SPD to IDLE
2. VNAV bugs for 250 knots as soon as the deceleration starts - before reaching 10.000 ft

**Comment by oMrSmith** [comment]
I think this is the solution: 
As soon as VNAV reduces the descend rate 
in order to decelerate down to 250 knots
it should bug for 250 knots.

**Comment by mSparks43** [comment]
> Should VNAV already bug for 250 when starting the deceleration?

speed in VNAV PTH in descent is being calculated as a deceleration rate per feet of descent.

**Comment by oMrSmith** [comment]
I mean the SPD REST of 250 below 10.000

**Comment by mSparks43** [comment]
> I mean the SPD REST of 250 below 10.000

I know, 
But the path that creates the ToD is a descent path that assumes speed reduces at fixed rate for every x feet of descent, thats where this issue comes from, because in this context ToD is meaningless, what is needed to hit the 240/10,000 is some form of "start of deceleration", which isn't something I've seen in any of the documentation.

**Comment by oMrSmith** [comment]
Maybe there is no conflict between your point and my point:
I'll I'm saying is this: If the plane has crossed the deceleration point
"start of deceleration" it won't decelerate if VNAV wont bug for 240
and the thrust mode holds the CRZ spd.

![image](https://user-images.githubusercontent.com/78870770/118466591-4a0bbc00-b703-11eb-92cb-155325763db4.png)

**Comment by mSparks43** [comment]
> VNAV should bug for 240 already.

SPD changes gradually with altitude in VNAV descent along the path. Therefore/currently, a TOD at 11,000 feet with a 240/10,000 restriction assumes you are at about 250kts at the TOD.

coming into the restriction it then scales from current speed to the target speed during the descent (to keep the descent stable):
https://github.com/mSparks43/747-400/blob/ce3b4188a2461aea214e2270e25090978c31dd1b/plugins/xtlua/scripts/B747.70.xt.autopilot/B747.70.xt.autopilot.vnavspd.lua#L286

which will blow you well past the TOD if you reach it at 327kts because it wont have time to slow down, and will hold at around 10,000 feet bleeding off all the speed when you should have been descending.

**Comment by oMrSmith** [comment]
I think we might missanderstand each other :D
I'm not saying that the plane should automativcally
decellarate perfectly from cruzing speed to 240 within 1000 ft of descend! It's impossible.
I'm also not saying that the plane should be at the speed of 240 knots
when the decelleration begins.
All I'm saying is the thrust mode SPD ist killing the 
deceleration if the speed bug won't swtich to the new target of 240 knots soon enough.
I'll test this in the Zibo later this day and see how it works there...

**Comment by mSparks43** [comment]
> All I'm saying is, the thrust mode SPD **_is killing_** the
> deceleration

It should dial in to 240kts _after_ descending 500feet.
All the transitions here are staged so each change is stable before the next change.

**Comment by crazytimtimtim** [comment]
> https://github.com/mSparks43/747-400/blob/ce3b4188a2461aea214e2270e25090978c31dd1b/plugins/xtlua/scripts/B747.70.xt.autopilot/B747.70.xt.autopilot.vnavspd.lua#L286

What does `B747_rescale` do?

**Comment by mSparks43** [comment]
> What does `B747_rescale` do?

changes one range into another given a current value
here lower alt will be 10000 feet, spdval will be 270kts and nextspeedval will be 240kts

and it will return 270kts at 11500 feet and above
255kts at 11250 feet
and 240kts at 11000 feet and below

**Comment by oMrSmith** [comment]
The issue is not a biggy for me. If VNAV refuses to bug for 240 immediately, I'll open the speed-window and do it myself.
However: I stand by my word.
It works the same in the 747 by SSG and in the 737 by ZIBO. I just tested it:
https://youtu.be/ZQkW9k5oD0k

**Comment by crazytimtimtim** [comment]
> > What does `B747_rescale` do?
> 
> changes one range into another given a current value
> here lower alt will be 10000 feet, spdval will be 270kts and nextspeedval will be 240kts
> 
> and it will return 270kts at 11500 feet and above
> 255kts at 11250 feet
> and 240kts at 11000 feet and below

Got it, thanks.

**Comment by mSparks43** [comment]
> 2\. VNAV bugs for 250 knots as soon as the deceleration starts - before reaching 10.000 ft

the problem with dropping speed so much from current, is if you engage FLCH at this point you will climb a fair bit - maybe even a lot to lose the speed before the descent resumes.

**Comment by oMrSmith** [comment]
_if you engage FLCH at this point you will climb a fair bit_
I don't think so. FLC does not work like IAS HOLD,
it will not climb when the selected MCP ALT is lower
and not descend when the selected MCP ALT is higher,
it will _level off_.

**Comment by mSparks43** [comment]
> it will not climb if the selected MCP ALT is lower

It definitely does climb if IAS is significantly lower that current speed, and also sink if significantly higher than current speed.

The first thing it does is match speed by pitching up until the speed is reached when lower or pitching down until the speed is reached when higher (regardless of MCP ALT), then once it it settles begins to descend or climb in the correct direction. 
One of the reasons I added the increments is VNAV SPD in climb would often lose up to 2000 feet in a 180->250kt conversion.

**Comment by oMrSmith** [comment]
Although the selected MCP ALT is lower "It definitely does climb if IAS is significantly lower that current speed"
I bet 50 bucks it doesn't! :D

**Comment by mSparks43** [comment]
> I bet 50 bucks it doesn't! :D
If your current speed is lower than that bugged, and you convert to FLCH, then convert to a speed slower than current speed it overcompensates (and vice versa)- potentially by going several thousand feet in the wrong direction depending upon the circumstance and timing.

It looks like it overcompensates more if you want to climb than descend. But its definitely in both directions
![image](https://user-images.githubusercontent.com/6773686/118668168-ea8dd900-b7f4-11eb-917d-cc71d1bd2518.png)

**Comment by oMrSmith** [comment]
FLC does not climb one foot. ;)
https://youtu.be/WMRaeMia7hw

**Comment by mSparks43** [comment]
> FLC does not climb one foot. ;)

_should_ not climb one foot, but if the speed bug is changed rapidly in _unstable_ flight it does (took me a few goes spinning the bug)
https://youtu.be/exTSazCT_LM

Because the nose ends up pitched up enough to create a climb/descent in the wrong direction before it registers that pitch will create a climb/descent in the wrong direction and it then takes multiple seconds to correct it (and in the meantime its climbing/descending in the wrong direction)

Alternate fix to this (rather than graduating the speed) is to speed up the rate it can change pitch - however then approach ends up unstable with pitch changing to quickly when it captures the glideslope.

**Comment by oMrSmith** [comment]
Oh yeah, I agree. But it's not something the AP intends to do, rather a normal side effect...

**Comment by mSparks43** [comment]
> rather a normal side effect...

It's hard to do with the dial by hand, because it takes several seconds to dial in a new speed, but very easy to make happen if VNAV snaps the speed directly in big jumps - i.e. if

> 2\. VNAV bugs for 250 knots as soon as the deceleration starts - before reaching 10.000 ft

This is why all the changes in VNAV ended up staged and graduated, speed change generally limited between 5-20kts in any one go, rather than having multiple targets all not being met at the same time by large amounts - which would demand large swings in pitch ("OMFG I'm 100kts faster than I'm supposed to be, must slam on the brakes now").

**Comment by oMrSmith** [comment]
Hmm. But VNAV PTH does not pitch for speed like VNAV SPD or FLC.
So in case of the SPD REST of 240/10000 bugging for 240
_before_ reaching 10000 feet just makes one litte but significant differance:
The thrust will go IDLE so that the speed will slowly drop, don't you think?

**Comment by mSparks43** [comment]
> _before_ reaching 10000 feet

but thats my point about some kind of "deceleration point". 

It does, starting from 1500 feet before 10,000 feet, with VNAV path assuming you have already slowed down to the 270kt descent speed, but only once you are 500 feet below cruise altitude in a stable descent.

**Comment by oMrSmith** [comment]
I'll check it out now!

**Comment by oMrSmith** [comment]
You know what is great? This plane can now fly RNAV approaches. My 2. test on LXGB went very well.

**Comment by oMrSmith** [comment]
The speed-big thing worked similar than before I think. However: As the plane would bug down shortly after T/D the thrust went IDLE and the deceleration could take effect. Maybe we'll leave it as it is for now... There's one last question: I understand that the SPD bug goes down gradually like you have explained. However there is an exception: When the plane reaches 10.000 all of a sudden the speed-bug jumps to 240 knots. If that exception would happen early (as soon as VNAV PTH starts to reduce the descend rate to around 500 fpm) it would work like in the 747 SSG or Zibo 737.

**Comment by mSparks43** [comment]
so what i changed was it was engaging the autothrottle in that zone, this would put throttle on to accelerate to the bug speed. what i changed was locking that out until actual speed is close to the restriction speed (275kts above 10k 245 below).
when it "jumps" to 240kts, throttle should be idle and it will/should be in a stable 500fpm decelerated vs descent. coming from a realistic cruise altitude (FL200 and above), normal graduation applies, restriction alt +1000 bugs to the restriction speed and follows #518

**Comment by oMrSmith** [comment]
_"When it "jumps" to 240kts, throttle should be idle and it will/should be in a stable 500fpm decelerated vs descent."_

**Exactly!** But if this (speed-bug) JUMP happens at 10.000 ft, it is too late, as the deceleration should already by acomplished by then!

at the _"restriction alt +1000"_ it _"bugs to the restriction speed"_
That's what I think too! But the plane bugs for 240 at 10.000 ft and not at 11.000 ft.

![image](https://user-images.githubusercontent.com/78870770/118951309-47a39f00-b95b-11eb-9f87-b65bcf818112.png)

**Comment by crazytimtimtim** [comment]
> and 240kts at 11000 feet and below

interesting, mSparks said this ^ earlier.
So I think he doesn't understand you because he thinks it is already done in the Sparky, but in reality, just doesn't work properly.

**Comment by mSparks43** [comment]
> So I think he doesn't understand you because he thinks it is already done in the Sparky,

I understand, the speed bug is irrelevant (not used/ignored; 500fpm descent and idle throttle start at ToD, what the speedbug is set to at that point [no longer] matters)
The issue is at 11,000 feet and 340kts there simply isn't time to decelerate to 240kts in the two minutes it takes to descend 1000 feet at 500fpm

**Comment by oMrSmith** [comment]
"the speed bug is irrelevant" I don't think so:
It has a significant impact on the active thrust mode.

![image](https://user-images.githubusercontent.com/78870770/119000473-6de13300-b98b-11eb-9b1b-2a10d385e660.png)

If the speed bug would change to 240 soon enough
the thrust mode would be "IDLE" and not hold the CRZ SPD.

**Comment by mSparks43** [comment]
> _It has a significant impact on the active thrust mode._

this is what changed in https://github.com/mSparks43/747-400/commit/79b13673dada7b33f329b660e6cd0cf360733b91
^^
previously SPD would always engage regardless of current speed, now 
```Lua
and simDR_ind_airspeed_kts_pilot < spdval+5 
```

**Comment by oMrSmith** [comment]
@mSparks43 I have an Idea for a possible compromise: 
Do you think you could add a fix 
"IDLE thrust after passing ToD even if CRZ ALT is low (like 11.000)"?

**Comment by mSparks43** [comment]
> "IDLE thrust after passing ToD even if CRZ ALT is low (like 11.000)"?

thing is, cruise alt is +/-500 feet to account for pressure changes, that's one of the reasons why its not spinning down or dropping the thrust until 10,500 feet, changing speed target _and_ VS at the same time makes the descent fairly unstable.

The issue here is almost identical to trying to convert from cruise speed at 10,000 feet to a 240kt restriction, the only difference is there is a 500ft high window to try and slow down, but that still isn't enough.

I have considered reversing them (so convert to descent thrust before descent speed), but I dont think that will change much here, and is likely to make descent from a normal cruise alt fall short, also presents problems with when exactly to switch because then it would be holding an altitude with no thrust rapidly loosing speed for no (normal) reason.

We're going round in circles because we're guessing - I don't even know for sure 11,000 feet is a valid cruise altitude. The minimum would normally be FL330, and entirely possible cruise altitude _has_ to be above transition altitude (generally FL180)

**Comment by oMrSmith** [comment]
I think the issue would solve itself if the SPD BUG change would trigger sooner. But you're the boss. We'll leave it like it is then and I'll close the issue.

**Comment by crazytimtimtim** [comment]
I see, so the issue is with his cruise alt of 11.000ft?

**Comment by mSparks43** [comment]
> I see, so the issue is with his cruise alt of 11.000ft?

The issue is with decelerating from 327kts to 240kts between 11,000 and 10,000 feet while also descending at 500fpm.

There is an issue here (VNAV is trying to do it and can't succeed)
But the resolution isn't clear - should it do something differently, should it refuse try, is this to be expected and the pilot should begin deceleration early.

about the only thing close I've found in the FCOM is an FMC page we dont have yet
![image](https://user-images.githubusercontent.com/6773686/119023873-bfe18300-b9a2-11eb-8abe-32d38cd8ea8c.png)

The drag required message can also be smarter to deal with this situation.

**Comment by crazytimtimtim** [comment]
I think it tries, it'll give you DRAG REQUIRED, but it'll just keep going and will decelerate once at 10,000 feet. But idk. 

I think when entering the flight plan there are a few warnings that can show up such as UNABLE TARGET AT [WAYPOINT NAME], but it'll continue anyway.

**Comment by crazytimtimtim** [comment]
> I just would expect the plane to bug for the target speed as soon as the plane
> _tries_ to decelerate to that speed as every boeing add on that I know
> does exactly that https://youtu.be/ZQkW9k5oD0k - causing the thrust to go idle so that the reduced descend rate
> slows the plane down... even if its not enough to meet the target speed when reaching 10.000 feet.
> The rest is up to the pilot: Level off, speed brakes...

I agree with this.

**Comment by mSparks43** [comment]
> I just would expect the plane to bug for the target speed as soon as the plane
> _tries_ to decelerate to that speed as every boeing add on that I know

VNAV here is custom and matched to the FCOM - can't speak for the others, XPLANE offers default VNAV but it is very not to the Boeing FCOM.

Level Deceleration starts as soon as the plane crosses ToD
![image](https://user-images.githubusercontent.com/78870770/118951309-47a39f00-b95b-11eb-9f87-b65bcf818112.png)
500 feet below cruise alt it transitions from 2 to 3
![image](https://user-images.githubusercontent.com/6773686/119035887-5ec0ac00-b9b0-11eb-8ff3-0cb855115ba6.png)

BUT.
The descent path at 10,500 feet assumes you are doing 240kts, flying at 11,000feet you cut out 3 altogether and jump straight from 2 to 4 at 10,500 feet.

**Comment by crazytimtimtim** [comment]
You couldn't do level decel at 11,000 ft?

**Comment by mSparks43** [comment]
> You couldn't do level decel at 11,000 ft?

no, because there's no guarantee the weather/turbulence didn't result in not being at exactly 11,000 - I put a +/- 500 foot window around the cruise altitude which count as "cruise alt".

however looking at that figure again, 
>Or
when flying so low with a speed restriction near and just below ToD the ToD is increased automatically to account for the need to burn off speed much faster than is normal

Sounds a lot like the flat line between ToD and the descent elbo.

**Comment by crazytimtimtim** [comment]
> I put a +/- 500 foot window around the cruise altitude

tbh I think 500ft is a bit much. On the Zibo, for example, it always holds altitude within 100 feet of my setting. Never flown irl though.

**Comment by mSparks43** [comment]
> tbh I think 500ft is a bit much.

https://www.youtube.com/watch?v=mogJHXicqhc

**Comment by oMrSmith** [comment]
Hey @mSparks43 with the following video
I want to share some general feedback 
on how the AP has developed lately:
https://youtu.be/w3PoTnkLfxE?t=48
